File: practice/src/test/java/com/decade/practice/web/SecurityFilterTest.java
package com.decade.practice.web;

import com.decade.practice.common.BaseTestClass;
import com.decade.practice.infra.security.UserClaimsTokenService;
import com.decade.practice.infra.security.models.UserClaims;
import com.decade.practice.utils.TokenUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.web.servlet.MockMvc;

import java.util.UUID;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@Sql(scripts = "/sql/clean.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
class SecurityFilterTest extends BaseTestClass {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private UserClaimsTokenService tokenService;

    @Test
    @Sql(scripts = {"/sql/clean.sql", "/sql/seed_users.sql"})
    void givenUnauthenticatedUser_whenAccessProtectedResource_thenReturnsUnauthorized() throws Exception {
        // /accounts/me is protected
        mockMvc.perform(get("/accounts/me"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    @Sql(scripts = {"/sql/clean.sql", "/sql/seed_users.sql"})
    void givenUnauthenticatedUser_whenAccessPublicResource_thenReturnsOk() throws Exception {
        // /users is public for registration (POST), but GET might also be public or have different rules.
        // Let's check /medias/** or /users/** (registration)
        // Actually SecurityConfiguration says .requestMatchers("/users/**").permitAll()
        mockMvc.perform(get("/users").param("query", "alice"))
                .andExpect(status().isOk());
    }

    @Test
    @Sql(scripts = {"/sql/clean.sql", "/sql/seed_users.sql"})
    void givenValidJwtToken_whenAccessProtectedResource_thenReturnsOk() throws Exception {
        // Given
        UserClaims claims = UserClaims.builder()
                .id(UUID.fromString("11111111-1111-1111-1111-111111111111"))
                .username("alice")
                .name("alice")
                .role("ROLE_USER")
                .build();
        String token = tokenService.create(claims, null).getAccessToken();

        // When & Then
        mockMvc.perform(get("/accounts/me")
                        .header(TokenUtils.HEADER_NAME, TokenUtils.BEARER + " " + token))
                .andExpect(status().isOk());
    }

    @Test
    @Sql(scripts = {"/sql/clean.sql", "/sql/seed_users.sql"})
    void givenExpiredOrInvalidToken_whenAccessProtectedResource_thenReturnsUnauthorized() throws Exception {
        mockMvc.perform(get("/accounts/me")
                        .header(TokenUtils.HEADER_NAME, TokenUtils.BEARER + " invalid-token"))
                .andExpect(status().isUnauthorized());
    }
}

File: practice/src/test/java/com/decade/practice/authentication/SecurityFilterTest.kt
package com.decade.practice.authentication

import com.decade.practice.DevelopmentApplication
import com.decade.practice.MockEndpoints
import com.decade.practice.core.ChatOperations
import com.decade.practice.core.UserOperations
import com.decade.practice.database.repository.AdminRepository
import com.decade.practice.database.repository.UserRepository
import com.decade.practice.model.domain.entity.User
import com.decade.practice.model.local.Account
import com.decade.practice.security.DaoUserDetailsService
import com.decade.practice.security.SecurityConfiguration
import com.decade.practice.security.jwt.JwtCredentialService
import com.decade.practice.security.jwt.JwtTokenFilter
import com.decade.practice.security.strategy.LoginSuccessStrategy
import com.decade.practice.security.strategy.LogoutStrategy
import com.decade.practice.security.strategy.Oauth2LoginSuccessStrategy
import com.decade.practice.session.SessionConfiguration
import com.decade.practice.utils.DummyRedisSetOps
import com.decade.practice.utils.TokenUtils
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration
import org.springframework.boot.autoconfigure.session.SessionAutoConfiguration
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.system.OutputCaptureExtension
import org.springframework.context.annotation.Import
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestBuilders
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@WebMvcTest(controllers = [MockEndpoints::class])
@ActiveProfiles("development")
@ContextConfiguration(classes = [DevelopmentApplication::class])
@ExtendWith(OutputCaptureExtension::class)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@Import(
      SecurityConfiguration::class,
      SessionConfiguration::class,
      LoginSuccessStrategy::class,
      JwtTokenFilter::class,
      DaoUserDetailsService::class,
      MockEndpoints::class,
      RedisAutoConfiguration::class,
      Oauth2LoginSuccessStrategy::class,
      SessionAutoConfiguration::class,
      JwtCredentialService::class,
)
class SecurityFilterTest {

      @MockBean
      lateinit var userRepo: UserRepository

      @MockBean
      lateinit var adminRepository: AdminRepository

      @MockBean
      lateinit var chatOperations: ChatOperations

      @MockBean
      lateinit var userService: UserOperations

      @MockBean
      lateinit var logoutHandler: LogoutStrategy

      @Autowired
      lateinit var mockMvc: MockMvc

      @Autowired
      lateinit var encoder: PasswordEncoder

      @MockBean
      lateinit var redisTemplate: StringRedisTemplate

      @Autowired
      lateinit var credentialService: JwtCredentialService

      @BeforeEach
      fun setUp() {
            val user = User(USERNAME, encoder.encode(PASSWORD))
            Mockito.`when`(userRepo.getByUsername(USERNAME)).thenReturn(user)
            Mockito.`when`(userService.prepareAccount(Mockito.any<UserDetails>())).thenReturn(Account(user, null))
            Mockito.`when`(redisTemplate.opsForSet()).thenReturn(DummyRedisSetOps<String, String>())
      }

      @Test
      @Throws(Exception::class)
      fun given_nonExistentUsername_when_login_then_failsWithUnauthorized() {
            mockMvc.perform(SecurityMockMvcRequestBuilders.formLogin().user("vcl").password(PASSWORD))
                  .andExpect(MockMvcResultMatchers.status().isUnauthorized())
                  .andExpect(MockMvcResultMatchers.content().string("Username not found"))
      }

      @Test
      @Throws(Exception::class)
      fun given_validCredentials_when_login_then_succeeds() {
            mockMvc.perform(SecurityMockMvcRequestBuilders.formLogin().user(USERNAME).password(PASSWORD))
                  .andExpect(MockMvcResultMatchers.status().isOk())
      }

      @Test
      @Throws(Exception::class)
      fun given_validUsernameWithInvalidPassword_when_login_then_failsWithUnauthorized() {
            mockMvc.perform(
                  SecurityMockMvcRequestBuilders.formLogin()
                        .user(USERNAME)
                        .password("vcl")
            )
                  .andExpect(MockMvcResultMatchers.status().isUnauthorized())
                  .andExpect(MockMvcResultMatchers.content().string("Wrong password"))
      }

      @Test
      @Throws(Exception::class)
      fun given_validJwtToken_when_requestProtectedResource_then_succeeds() {
            val user = userRepo.getByUsername(USERNAME)
            val accessToken = credentialService.create(user).accessToken

            mockMvc.perform(
                  MockMvcRequestBuilders.get("/mock")
                        .header(TokenUtils.HEADER_NAME, "${TokenUtils.BEARER} $accessToken")
            )
                  .andExpect(MockMvcResultMatchers.status().isOk())
      }

      companion object {
            private const val USERNAME = "mock_username"
            private const val PASSWORD = "mock_password"
      }
}